### PR TITLE
runtime: remove backup_thread_running

### PR DESCRIPTION
**Pull request by gasche**
This is a proposed fix for #13677, that simply removes `backup_thread_running` and checks `backup_thread_msg` instead.

### CODE REVIEW

File: runtime/domain.c
Comment (gasche) [inline runtime/domain.c]: Note: in this function I don't understand why we are computing `backup_thread_running` unconditionally, even when `!caml_domain_alone` holds. I left the code as-is.

File: runtime/domain.c
Comment (gasche) [inline runtime/domain.c:2209]: Note: this looks like a code smell, we are busy-looping here when I suppose we should use locking primitives to wait for termination more gracefully, in the same way that we do in `install_backup_thread`. I wonder if @dustanddreams has an opinion -- this comes from his heroic work on #12964.

**Comment by jmid** [comment]
Thanks @gasche. Can you try labeling it with `run-multicoretests`? :pray:

File: runtime/domain.c
Comment (gasche) [inline runtime/domain.c:1349]: Note: the logic here feels a bit fragile to me, because in general a backup thread can be running that comes from an older domain (an older instance of the same domain slot) that is in the process of terminating. This should not happen in the case of domain 0 that we are thinking about here, but I don't see any robust check that this is in fact the first spawn from domain 0, so from a distance it looks like we may fail to register a backup thread for some newly spawned domain if a previous domain is not done with its own backup thread in the same slot.

**Review by gasche — COMMENTED** [review]
I made some self-review comments that are orthogonal to the change at hand (which dutifully follows the current code) but point out things I don't understand in the current implementation.

**Review by jmid — APPROVED** [review]
This change makes sense to me, based on having had my fingers a bit in these parts in #12707.
AFAICS, the backup thread logic has evolved over several iterations, and it does seem simpler and more robust to just represent their state as a single atomic value.

I also gave this a stress test spin on `multicoretests` where it holds up nicely:
https://github.com/ocaml-multicore/multicoretests/actions?query=branch%3Atarget-gasches-13679-incl-gctests
(the only failure is a debug assertion run from #13090)

Not considering myself a runtime expert, I would be more at ease if, e.g., @kayceesrk who authored backup threads initially could also take a look.

File: runtime/domain.c
Comment (ghost) [reply runtime/domain.c]: This is because the name of the function is misleading. What it really tests for is "is there more than one thread running?".

Suppose your program spawned N domains, and all these N domains have exited, `caml_domain_alone` will return `true`, but the backup thread for the main domain has nevertheless been started in the meantime, and is probably still running.

Hence the check for the BT here.

File: runtime/domain.c
Comment (ghost) [reply runtime/domain.c:2209]: When this particular code path is hit, we are in "cleanup at exit" mode, and want to ensure, while terminating a domain, that its backup thread terminates, before processing the next domain. In this case the backup thread has already been properly signaled, what really is needed is for it to be scheduled - either on the current processor or on another one. Using `cpu_relax` here should be acceptable because the loop really won't take long.

**Review by ghost — APPROVED** [review]
Looking good (not in a position to torture test this at the moment). It had already been pointed out that `backup_thread_running` was a debug relic, adding no real value, so these changes go in the Right Direction™.

File: runtime/domain.c
Comment (gasche) [reply runtime/domain.c]: Thanks! My question, more precisely, was why we are not taking advantage of the shortcut behavior of the `||` by writing `!caml_domain_alone || backup_thread_running(self)`.

Comment (ghost) [reply runtime/domain.c]: Ah!

The reason is that, prior to this PR, folding the backup thread check into the `if` would make it no longer fit in a single line. With your new wrapper routine, it ought to fit, so go for it!

File: runtime/domain.c
Comment (gasche) [reply runtime/domain.c:2209]: I'm not convinced (I would rather have a proper synchronization operation within the busy loop), but then I realized that we cannot do what the other loop does anyway, it releases and retakes the domain lock but here we are trying to terminate the backup thread of *other* domains whose domain lock we do not own. I'm leaving the code as-is.

File: runtime/domain.c
Comment (gasche) [reply runtime/domain.c]: Thanks, I made this small additional change.

**Comment by gasche** [comment]
The reviews of @jmid and @dustanddreams helped convinced me that this is a good idea -- thanks! Now this needs a maintainer approval to move forward. (Maybe @OlivierNicole or @nojb or @NickBarnes or @kayceesrk?)

**Comment by nojb** [comment]
> Now this needs a maintainer approval to move forward. (Maybe @OlivierNicole or @nojb or @NickBarnes or @kayceesrk?)

The intricacies of parallel programming are a bit over my head, so I cannot say that I am fully confident about the change (although it looks reasonable). But I am happy to approve on the strenght of the existing reviews by @jmid and @dustanddreams.

**Review by nojb — APPROVED** [review]
LGTM

Approved on behalf of @jmid and @dustanddreams.

**Pull request merged**
